### PR TITLE
[WIP] Module export using custom file format

### PR DESCRIPTION
**Pull request by jamesr66a**
Stacked on https://github.com/pytorch/pytorch/pull/9746. Look at the last commit.

This implements module (de)serialization using a custom aligned data format

### CODE REVIEW

File: torch/csrc/jit/export.cpp [hunk 1]
```diff
@@ -756,6 +757,174 @@ std::tuple<std::string, RawDataExportMap> ExportGraph(
...
+
+  size_t getModelProtoOffset() const {
+    return key_to_file_offset.at("__MODEL_PROTO");
+  }
+
+  // Serialize the model proto to file, replacing the tensor names with
+  // string verisions of their file offsets. This also finalizes the file,
+  // and calling serializeTensor after calling this method is illegal.
+  // NOTE: this method mutates the model proto
+  size_t serializeModelProto(::ONNX_NAMESPACE::ModelProto *model_proto) {
```
Comment (zdevito) [inline torch/csrc/jit/export.cpp]: This is not a good way to do this. It violates abstraction boundaries. Now this Reader/Writer are very coupled to the meta-data format, and can never be used to hold other data. Other data in the future may be: pickle data from Python, Cereal data from C++, etc.

This reader/writer should know nothing about ModelProto files.

File: torch/csrc/jit/export.cpp [hunk 2]
```diff
@@ -756,6 +757,174 @@ std::tuple<std::string, RawDataExportMap> ExportGraph(
...
 
+class PyTorchFileWriter {
+public:
+  PyTorchFileWriter(const std::string& filename){
+    fp = std::fopen(filename.c_str(), "wb");
+    writeFileHeader();
+  }
+
+  // Serialize a tensor to file, then return its offset
+  size_t serializeTensor(const std::string& name, at::Tensor t) {
```
Comment (zdevito) [inline torch/csrc/jit/export.cpp]: This should not take a name. It should take data and return an 8-byte key to where that data is. What the key is should be abstract in the API of this format.

File: torch/csrc/jit/export.cpp (hunk 1 above)
Comment (zdevito) [reply torch/csrc/jit/export.cpp]: This logic can instead be moved entirely into ModelEncoder and ModelDecoder.

File: torch/csrc/jit/export.cpp (hunk 2 above)
Comment (zdevito) [reply torch/csrc/jit/export.cpp]: Signature should be:
```uint64_t writeData(char* data, size_t size_in_bytes)```

This should not have to know what a tensor is. We will use it to store non-tensor data.

File: torch/csrc/jit/export.cpp [hunk 3]
```diff
@@ -756,6 +757,174 @@ std::tuple<std::string, RawDataExportMap> ExportGraph(
...
+    writePad(pad_amount);
+  }
+
+  void writeFileHeader() {
+    write64BitIntegerLittleEndian(kFileMagicNumber);
+    write64BitIntegerLittleEndian(kFileFormatVersion);
+    padToNextAlignmentBoundary();
+  }
+
+  void swapTensorAttributeNames(::ONNX_NAMESPACE::GraphProto *g) {
```
Comment (zdevito) [inline torch/csrc/jit/export.cpp]: This is very fragile and has huge action-at-a-distance problems. Please make an abstract interface that is used to write the tensors instead.

File: torch/csrc/jit/import.cpp [hunk 4]
```diff
@@ -319,4 +320,113 @@ std::shared_ptr<script::Module> ImportIRModule(
...
+ public:
+  PyTorchFileReader(const std::string& filename) {
+   fp = std::fopen(filename.c_str(), "rb");
+   std::fseek(fp, 0L, SEEK_END);
+   file_size = std::ftell(fp);
+   std::fseek(fp, 0L, SEEK_SET);
+   readAndValidateFileHeader();
+  }
+
+  // returns raw data map and the index of the last record (i.e. the ModelProto)
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: This isn't the API we eventually want. This PR should strive to get the API right for the reader or writer classes even if the implementation may change. Right now the API is weirdly designed to match up with the ModelDecoder API. We should be changing the API of both to create clear abstraction boundaries.

File: torch/csrc/jit/import.cpp [hunk 5]
```diff
@@ -319,4 +320,113 @@ std::shared_ptr<script::Module> ImportIRModule(
...
+  std::tuple<std::unordered_map<std::string, std::string>&, size_t> read_raw() {
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: `shared_ptr<void> readRecord(int64_t key)`

File: torch/csrc/jit/import.cpp [hunk 6]
```diff
@@ -319,4 +320,113 @@ std::shared_ptr<script::Module> ImportIRModule(
...
+      final_record_offset = readRecordIntoMap();
+    }
+    finalized = true;
+    return {raw_values, final_record_offset};
+  }
+
+  ~PyTorchFileReader() {
+    std::fclose(fp);
+  }
+
```
Comment (zdevito) [inline torch/csrc/jit/import.cpp]: `shared_ptr<void> readLastEntry()`

**Review by zdevito — COMMENTED** [review]
Generally the on-disk format looks right, but the APIs for reading/writing need serious cleanup. We need to get them right otherwise the ways we will try to improve them won't be possible.

File: torch/csrc/jit/import.cpp (hunk 6 above)
Comment (jamesr66a) [reply torch/csrc/jit/import.cpp]: I think I'm going to scrap the concept of assuming that the last entry is the proto and instead encode that as an offset in the file header

File: torch/csrc/jit/export.cpp [hunk 7]
```diff
@@ -756,6 +757,174 @@ std::tuple<std::string, RawDataExportMap> ExportGraph(
   return std::make_tuple(model_proto.SerializeAsString(), graph_encoder.get_raw_data_export_map());
 }
 
+class PyTorchFileWriter {
```
Comment (dzhulgakov) [inline torch/csrc/jit/export.cpp]: can we actually put it in a separate file too, with clean API?

File: torch/csrc/jit/script/init.cpp [hunk 8]
```diff
@@ -385,6 +385,15 @@ void initJitScriptBindings(PyObject* module) {
         return std::make_tuple(py::bytes(module), python_serialized_export_map);
       }, py::arg("onnx_opset_version")=0,
          py::arg("operator_export_type")=::torch::onnx::OperatorExportTypes::RAW)
+      .def("export_to_pytorch_file", [](const std::shared_ptr<Module> m, const std::string& filename,
+                                        int64_t onnx_opset_version,
```
Comment (dzhulgakov) [inline torch/csrc/jit/script/init.cpp]: why does it have onnx opset version? will it actually work without onnx export path?

**Comment by jamesr66a** [comment]
Starting new PR soon

**Pull request closed without merging**